File: jcms/src/main/java/com/ysh/jcms/datatypes2/data/block/CmsUrcb.java
package com.ysh.jcms.datatypes2.data.block;

import com.sun.jna.ptr.IntByReference;
import com.ysh.jcms.datatypes2.data.basic.CmsBitStringFixed;
import com.ysh.jcms.datatypes2.data.basic.CmsBoolean;
import com.ysh.jcms.datatypes2.data.basic.CmsInt16U;
import com.ysh.jcms.datatypes2.data.basic.CmsInt32U;
import com.ysh.jcms.datatypes2.data.basic.CmsOctetString;
import com.ysh.jcms.datatypes2.data.basic.CmsVisibleString;
import com.ysh.jcms.datatypes2.ffi.CmsFFI;
import com.ysh.jcms.datatypes2.ffi.CmsStructure;

import java.util.Arrays;
import java.util.List;

/**
 * cms_urcb_t — URCB (Unbuffered Report Control Block).
 */
public class CmsUrcb extends CmsStructure {
    public CmsVisibleString rptID = new CmsVisibleString(130);
    public CmsBoolean rptEna = new CmsBoolean();
    public CmsVisibleString datSet = new CmsVisibleString(256);
    public CmsInt32U confRev = new CmsInt32U();
    public CmsBitStringFixed optFlds = new CmsBitStringFixed(10);
    public CmsInt32U bufTm = new CmsInt32U();
    public CmsInt16U sqNum = new CmsInt16U();
    public CmsBitStringFixed trgOps = new CmsBitStringFixed(6);
    public CmsInt32U intgPd = new CmsInt32U();
    public CmsBoolean gi = new CmsBoolean();
    public CmsBoolean resv = new CmsBoolean();
    public CmsOctetString owner = new CmsOctetString(64);
    public CmsInt32U ownerLen = new CmsInt32U();
    public CmsBoolean ownerPresent = new CmsBoolean();

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("rptID", "rptEna", "datSet", "confRev", "optFlds", "bufTm",
                "sqNum", "trgOps", "intgPd", "gi", "resv", "owner", "ownerLen", "ownerPresent");
    }

    @Override protected int ffiEncode(byte[] buf, IntByReference outLen) { return CmsFFI.INSTANCE.cms_urcb_encode(this, buf, outLen); }
    @Override protected void ffiDecode(byte[] data) { CmsFFI.INSTANCE.cms_urcb_decode(data, data.length, this); }
    @Override protected int encodeBufSize() { return 512; }
    public static CmsUrcb from(byte[] data) { return new CmsUrcb().decode(data); }
}

File: ccms2/src/data/block/cms_urcb.c
#include "data/block/cms_urcb.h"
#include "data/string/cms_visible_string.h"
#include "data/string/cms_octet_string.h"

int cms_urcb_encode_stream(per_stream_t *s, const void *ptr) {
    const cms_urcb_t *pdu = (const cms_urcb_t*)ptr;
    int err;

    /* 1. rptID — VisibleString (SIZE(129)) */
    if (!pdu->rptID) return CMS_ERR;
    err = cms_visible_string_encode_stream(s, pdu->rptID, CMS_URCB_RPT_ID_MAX_LEN);
    if (err) return err;

    /* 2. rptEna — BOOLEAN */
    if (!pdu->rptEna) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->rptEna);
    if (err) return err;

    /* 3. datSet — ObjectReference */
    if (!pdu->datSet) return CMS_ERR;
    err = cms_object_reference_encode_stream(s, pdu->datSet);
    if (err) return err;

    /* 4. confRev — INT32U */
    if (!pdu->confRev) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->confRev);
    if (err) return err;

    /* 5. optFlds — RCBOptFlds */
    if (!pdu->optFlds) return CMS_ERR;
    err = cms_rcb_opt_flds_encode_stream(s, pdu->optFlds);
    if (err) return err;

    /* 6. bufTm — INT32U */
    if (!pdu->bufTm) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->bufTm);
    if (err) return err;

    /* 7. sqNum — INT16U */
    if (!pdu->sqNum) return CMS_ERR;
    err = cms_int16u_encode_stream(s, pdu->sqNum);
    if (err) return err;

    /* 8. trgOps — TriggerConditions */
    if (!pdu->trgOps) return CMS_ERR;
    err = cms_trigger_conditions_encode_stream(s, pdu->trgOps);
    if (err) return err;

    /* 9. intgPd — INT32U */
    if (!pdu->intgPd) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->intgPd);
    if (err) return err;

    /* 10. gi — BOOLEAN */
    if (!pdu->gi) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->gi);
    if (err) return err;

    /* 11. resv — BOOLEAN [14] */
    if (!pdu->resv) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->resv);
    if (err) return err;

    /* 12. owner — OCTET STRING (SIZE(0..64)) OPTIONAL */
    {
        int present = (pdu->owner_present && pdu->owner_present->value) && pdu->owner;
        cms_boolean_t bit = { .value = present };
        err = cms_boolean_encode_stream(s, &bit);
        if (err) return err;
        if (present) {
            err = cms_octet_string_encode_stream(s, pdu->owner, CMS_URCB_OWNER_MAX_LEN);
            if (err) return err;
        }
    }

    return CMS_OK;
}

int cms_urcb_decode_stream(per_stream_t *s, void *ptr) {
    cms_urcb_t *pdu = (cms_urcb_t*)ptr;
    int err;

    /* 1. rptID */
    if (!pdu->rptID) return CMS_ERR;
    err = cms_visible_string_decode_stream(s, pdu->rptID, CMS_URCB_RPT_ID_MAX_LEN);
    if (err) return err;

    /* 2. rptEna */
    if (!pdu->rptEna) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu->rptEna);
    if (err) return err;

    /* 3. datSet */
    if (!pdu->datSet) return CMS_ERR;
    err = cms_object_reference_decode_stream(s, pdu->datSet);
    if (err) return err;

    /* 4. confRev */
    if (!pdu->confRev) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu->confRev);
    if (err) return err;

    /* 5. optFlds */
    if (!pdu->optFlds) return CMS_ERR;
    err = cms_rcb_opt_flds_decode_stream(s, pdu->optFlds);
    if (err) return err;

    /* 6. bufTm */
    if (!pdu->bufTm) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu->bufTm);
    if (err) return err;

    /* 7. sqNum */
    if (!pdu->sqNum) return CMS_ERR;
    err = cms_int16u_decode_stream(s, pdu->sqNum);
    if (err) return err;

    /* 8. trgOps */
    if (!pdu->trgOps) return CMS_ERR;
    err = cms_trigger_conditions_decode_stream(s, pdu->trgOps);
    if (err) return err;

    /* 9. intgPd */
    if (!pdu->intgPd) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu->intgPd);
    if (err) return err;

    /* 10. gi */
    if (!pdu->gi) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu->gi);
    if (err) return err;

    /* 11. resv */
    if (!pdu->resv) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu->resv);
    if (err) return err;

    /* 12. owner — OCTET STRING OPTIONAL */
    {
        cms_boolean_t bit = {0};
        err = cms_boolean_decode_stream(s, &bit);
        if (err) return err;
        if (bit.value && pdu->owner) {
            err = cms_octet_string_decode_stream(s, pdu->owner, CMS_URCB_OWNER_MAX_LEN);
            if (err) return err;
        }
        if (pdu->owner_present) pdu->owner_present->value = bit.value;
    }

    return CMS_OK;
}

int cms_urcb_encode(const void *ptr, uint8_t *out_buf, int *out_len) {
    per_stream_t s;
    per_stream_init_write(&s, out_buf, (size_t)*out_len);
    int rc = cms_urcb_encode_stream(&s, ptr);
    if (rc) return rc;
    *out_len = (int)per_stream_bytes_written(&s);
    return CMS_OK;
}

int cms_urcb_decode(void *ptr, const uint8_t *in_buf, int in_len) {
    per_stream_t s;
    per_stream_init_read(&s, in_buf, (size_t)in_len);
    return cms_urcb_decode_stream(&s, ptr);
}
